python/guide/module/autograd.md

```diff
@@ -1,178 +1,286 @@
+
+What's the solution? Use automatic differenciation to backpropagate the gradients from the loss metric back to each of the parameters. Sometimes this is called reverse mode automatic differenciation, and it's very efficient in 'fan-in' situations where many parameters effect a single loss metric. Although forward mode automatic differenciation methods exist, they're suited to 'fan-out' situations where few parameters effect many metrics, which isn't the case for training neural networks.
+
+## How does Automatic Differenciation (`autograd`) work?
+
+Stage 1. Create a record of the operators used by the network to make predictions and calculate the loss metric. Called the 'forward pass' of training.
+Stage 2. Work backwards through this record and evaluate the partial derivatives of each operator, all the way back to the network parameters. Called the 'backward pass' of training.
+
+<p align="center">
+    <video width="400" controls playsinline autoplay muted loop>
+        <source src="./imgs/autograd_graph.mp4" type="video/mp4">
+    </video>
+</p>
+
+Long Answer:
+
+All operators in MXNet have two methods defined: a `forward` method for executing the operator as expected, and a `backward` method that returns the partial derivative (the derivative of the output with respect to the input). On the vary rare occasion you need to implement your own custom operator, you'll define the same two methods.
+
+Automatic Differenciation creates a record of the operators used (i.e. the `forward` method calls) by the network to make predictions and calculate the loss metric. A graph structure is used to record this, capturing the inputs (including their value) and outputs for each operator and how the operators are related. We call this the 'forward pass' of training.
+
+Automatic Differenciation then works backwards through each operator of the graph, calling the `backward` method on each operator to calculate the partial derivative and calculate the gradient of the loss metric with respect to the operator's input (which could be parameters). Usually we work backwards from the loss metric, and hence calculate the gradients of the loss metric, but this can be done from any output. We call this the 'backward pass' of training.
+
+### What are the advantages of Automatic Differenciation (`autograd`)?
+
+It's flexible, automatic and efficient. You can use native Python control flow operators such as `if` conditions and `while` loops and `autograd` will still be able to backpropogate the gradients correctly.
```

Comment:
I would add a link here to the chapter below: "Advanced: Using Python control flow"

Reply:
updated

Reply:
couldn't seem to link to internal section with colon in title.